File: service/src/test/java/cleveres/tricky/cleverestech/util/CborEncoderTest.java
package cleveres.tricky.cleverestech.util;

import org.junit.Test;
import java.util.HashMap;
import java.util.Map;
import static org.junit.Assert.assertArrayEquals;

public class CborEncoderTest {

    private String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02X", b));
        }
        return sb.toString();
    }

    @Test
    public void testMapSortingOrder() {
        // Map { 1: "a", -1: "b" }
        // CBOR:
        // 1 (unsigned) -> 0x01
        // -1 (negative) -> 0x20
        // Expected order: 1 (Major Type 0) comes before -1 (Major Type 1)

        Map<Integer, String> map = new HashMap<>();
        map.put(1, "a");
        map.put(-1, "b");

        byte[] encoded = CborEncoder.encode(map);

        // Expected:
        // A2       (Map, size 2)
        // 01       (Key: 1)
        // 61 61    (Val: "a")
        // 20       (Key: -1)
        // 61 62    (Val: "b")
        byte[] expected = new byte[] {
            (byte)0xA2,
            (byte)0x01, (byte)0x61, (byte)0x61,
            (byte)0x20, (byte)0x61, (byte)0x62
        };

        // If bug exists (Integer.compare used):
        // -1 comes before 1
        // A2
        // 20       (Key: -1)
        // 61 62    (Val: "b")
        // 01       (Key: 1)
        // 61 61    (Val: "a")

        System.out.println("Encoded: " + bytesToHex(encoded));
        assertArrayEquals("CBOR Map sorting is incorrect!", expected, encoded);
    }

    @Test
    public void testStringMapSortingOrder() {
        // Map { "aa": 1, "a": 2, "b": 3 }
        // Expected sort: "a", "b", "aa" (shorter length first)

        Map<String, Integer> map = new HashMap<>();
        map.put("aa", 1);
        map.put("a", 2);
        map.put("b", 3);

        byte[] encoded = CborEncoder.encode(map);

        // Expected:
        // A3       (Map 3)
        // 61 61    ("a")
        // 02       (2)
        // 61 62    ("b")
        // 03       (3)
        // 62 61 61 ("aa")
        // 01       (1)

        byte[] expected = new byte[] {
            (byte)0xA3,
            (byte)0x61, (byte)0x61, (byte)0x02,
            (byte)0x61, (byte)0x62, (byte)0x03,
            (byte)0x62, (byte)0x61, (byte)0x61, (byte)0x01
        };

        System.out.println("Encoded Strings: " + bytesToHex(encoded));
        assertArrayEquals("CBOR String Map sorting is incorrect!", expected, encoded);
    }
}

File: service/src/test/java/cleveres/tricky/cleverestech/CborEncoderTest.kt
package cleveres.tricky.cleverestech

import cleveres.tricky.cleverestech.util.CborEncoder
import org.junit.Test
import java.util.LinkedHashMap

class CborEncoderTest {

    private fun bytesToHex(bytes: ByteArray): String {
        val sb = StringBuilder()
        for (b in bytes) {
            sb.append(String.format("%02x", b))
        }
        return sb.toString()
    }

    @Test
    fun testCanonicalMapSorting() {
        // "b" (len 1) vs "aa" (len 2)
        // Canonical: "b" < "aa"
        // Java String: "aa" < "b"

        val map = LinkedHashMap<String, Int>()
        map["aa"] = 1
        map["b"] = 2

        // Expected Order: "b", "aa"
        // Encoded:
        // Map(2) (0xa2)
        // "b" (0x61 62) -> 2 (0x02)
        // "aa" (0x62 61 61) -> 1 (0x01)
        // Hex: a2 61 62 02 62 61 61 01

        val expected = "a261620262616101"
        val encoded = CborEncoder.encode(map)
        val encodedHex = bytesToHex(encoded)

        println("Encoded: $encodedHex")

        org.junit.Assert.assertEquals("Expected canonical sort order", expected, encodedHex)
    }

    @Test
    fun testCanonicalMapSortingLongerKeys() {
        // "manufacturer" (12) vs "vb_state" (8)
        // Canonical: "vb_state" < "manufacturer"

        val map = LinkedHashMap<String, Int>()
        map["manufacturer"] = 1
        map["vb_state"] = 2

        // Encoded:
        // Map(2) (0xa2)
        // "vb_state" (8 bytes) -> 0x68 + "vb_state" -> 2
        // "manufacturer" (12 bytes) -> 0x6c + "manufacturer" -> 1

        // 68 76625f7374617465 -> vb_state
        // 02
        // 6c 6d616e756661637475726572 -> manufacturer
        // 01

        val expected = "a26876625f7374617465026c6d616e75666163747572657201"
        val encoded = CborEncoder.encode(map)
        val encodedHex = bytesToHex(encoded)

        println("Encoded: $encodedHex")

        org.junit.Assert.assertEquals("Expected shorter key (vb_state) first", expected, encodedHex)
    }
}
